spec/index.md

```diff
@@ -4391,28 +4391,39 @@ Env = {
 }
 
 (
-  (A.mode = upgrade or A.mode = upgrade {skip_pre_upgrade = false})
+  (A.mode = upgrade U and U.skip_pre_upgrade ≠ true)
   Env1 = Env with {
     global_timer = S.global_timer[A.canister_id];
     canister_version = S.canister_version[A.canister_id];
   }
-  Old_module.pre_upgrade(Old_State, M.caller, Env1) = Return {stable_memory = Stable_memory; new_certified_data = New_certified_data; cycles_used = Cycles_used;}
+  Old_module.pre_upgrade(Old_State, M.caller, Env1) = Return {new_state = Intermediate_state; new_certified_data = New_certified_data; cycles_used = Cycles_used;}
 )
 or
 (
-  A.mode = upgrade {skip_pre_upgrade = true}
-  Stable_memory = Old_State.stable_mem
+  (A.mode = upgrade U and U.skip_pre_upgrade = true)
+  Intermediate_state = Old_state
   New_certified_data = NoCertifiedData
   Cycles_used = 0
 )
 
+(
+  (A.mode = upgrade U and U.wasm_memory_persistence ≠ keep)
+  Persisted_state = {store = initial_wasm_store; self_id = A.canister_id; stable_mem = Intermediate_state.stable_memory}
+)
+or
+(
+  (A.mode = upgrade U and U.wasm_memory_persistence = keep)
+  Persisted_state = Intermediate_state
```

Comment:
```suggestion
  Persisted_state = initialize_stote(Intermediate_state, A.wasm_module)
```
where `initialize_store(State, Wasm_module)` is the store of the WebAssembly module after *instantiation* (as per WebAssembly spec) of the WebAssembly module `Wasm_module` reusing the WebAssembly memory of `State`.

Reply:
Done without commenting on active data segments. @crusso Could you please comment on whether special treatment of active data segments is needed after all?

Reply:
@luc-blaeser and I discussed this offline and perhaps the simplest to specify is to disallow active data segments for EOP binaries - i.e, when keep is true. The motoko compiler already only emits passive data segments, rewriting the active data segment of the RTS to a passive one off-line.

Once could relax the restriction of no active data segments later  on in the spec and implementation.

Reply:
> disallow active data segments for (replacement) EOP (only) binaries - i.e, when keep is true

Why would it be simpler to disallow them rather than not commenting on them at all (meaning if the new module provides active data segments, then they'll overwrite parts of the persisted Wasm memory)?

Reply:
I had a corresponding prototype implementation, but there was a concern on complexity (and performance) of the upgrade that we have to selectively patch the active data segments. I believe a reason is that for the initial memory setup, the data segments can be applied in a simple page-based granularity (`Segments::as_pages()` extending the data segments to pages with zero initialization of the unused parts), while for the upgrade, the patching would need to be happen on the existing pages in the more fine-granular byte-wise approach.

The main reason was however that it is not very useful because the runtime system would need to reserve a fix section for active data segments in order not accidentally overwrite the heap. Choosing the size of the reserved space would be very difficult, it is often too wasteful or too limited. This is why we only use passive data segments, as it gives us the flexibility to choose the addresses dynamically in a simple way.